src/mini_batch.jl

```diff
@@ -0,0 +1,129 @@
+"""
+    MiniBatch(b::Int)
+
+    Sculley et al. 2007 Mini batch k-means algorithm implementation.
+"""
+struct MiniBatch <: AbstractKMeansAlg
+    b::Int  # batch size
+end
+
+
+MiniBatch() = MiniBatch(100)
+
+function kmeans!(alg::MiniBatch, X, k;
+                 weights = nothing, metric = Euclidean(), n_threads = Threads.nthreads(),
+                 k_init = "k-means++", init = nothing, max_iters = 300,
+                 tol = eltype(X)(1e-6), max_no_improvement = 10, verbose = false, rng = Random.GLOBAL_RNG)
+
+    # Get the type and dimensions of design matrix, X
+    T = eltype(X)
+    nrow, ncol = size(X)
+
+    # Initiate cluster centers - (Step 2) in paper
+    centroids = isnothing(init) ? smart_init(X, k, n_threads, weights, rng, init = k_init).centroids : deepcopy(init)
+
+    # Initialize counter for the no. of data in each cluster - (Step 3) in paper
+    N = zeros(T, k)
+
+    # Initialize nearest centers
+    labels = Vector{Int}(undef, alg.b)
+    final_labels = Vector{Int}(undef, ncol)
+
+    converged = false
+    niters = 0
+    counter = 0
+    J_previous = zero(T)
+    J = zero(T)
+
+    # TODO: Main Steps. Batch update centroids until convergence
+    while niters <= max_iters
+
+        # b examples picked randomly from X (Step 5 in paper)
+        batch_rand_idx = isnothing(weights) ? rand(rng, 1:ncol, alg.b) : wsample(rng, 1:ncol, weights, alg.b)
+        batch_sample = X[:, batch_rand_idx]
+
+        # Cache/label the batch samples nearest to the centers (Step 6 & 7)
+        @inbounds for i in axes(batch_sample, 2)
+            min_dist = distance(metric, batch_sample, centroids, i, 1)
+            label = 1
+
+            for j in 2:size(centroids, 2)
+                dist = distance(metric, batch_sample, centroids, i, j)
+                label = dist < min_dist ? j : label
+                min_dist = dist < min_dist ? dist : min_dist
+            end
+
+            labels[i] = label
```

Comment:
This is wrong, since labels is a vector of the whole `X` set, but you are updating indices `1:batch_size`. It should be something like `labels[batch_rand_idx[i]] = label`. And you should check other indices inside this procedure.

Reply:
Fixed